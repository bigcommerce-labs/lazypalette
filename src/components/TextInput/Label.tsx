import React from 'react';

import StyledLabel from './styles/StyledLabel';
import StyledSmall from './styles/StyledSmall';

interface LabelProps {
    htmlFor?: string;
    note?: string;
    required?: boolean;
}

class Label extends React.PureComponent<LabelProps> {
    render() {
        const { children: label, htmlFor, note, required } = this.props;

        return (
            <StyledLabel htmlFor={htmlFor}>
                {label}
                {note &&
                    <StyledSmall required={required!}>
                        {note}
                    </StyledSmall>
                }
            </StyledLabel>
        );
    }
}

export default Label;
