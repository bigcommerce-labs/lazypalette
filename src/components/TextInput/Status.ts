export enum Status {
  Invalid,
  Valid,
  Undefined,
}
