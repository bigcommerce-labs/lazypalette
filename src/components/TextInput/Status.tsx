export const enum Status {
    Invalid,
    Valid,
    Undefined,
}
