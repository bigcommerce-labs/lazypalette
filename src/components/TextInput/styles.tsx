import { theme } from 'pattern-lab';
import { Status } from './Status';
import styled, { StyledComponentClass } from 'styled-components';
import { colors } from '../../styleConstants';

const { colors: patternLabColors } = theme;
const defaultBorderColor = colors.lightBlueMagentaishGray;
const defaultBorderColorOnHover = colors.darkBlueMagentaishGray;
const immutableBorderColorOnFocus = 'none';
const immutableBorderColorOnHover = 'none';
const immutableInputBackgroundColor = colors.cyanishWhite;
const mutableInputBackgroundColor = colors.white;
const statuslessImmutableBorderColor = colors.cyanBluishWhite;
const warningBorderColor = colors.lightBrilliantOrangeYellow;

export const Container = styled.div`
    display: flex;
    flex-direction: column;
`;

export const Label = styled.label`
    padding: 5px;
`;

interface SmallProps {
    required: boolean;
}

export const Small = styled.small`
    float: right;
    color: ${(props: SmallProps) => props.required ? colors.lightBrilliantRedOrange : 'initial'};
`;

const getBackgroundColor = (props: TextInputProps) => {
    if (props.readonly || props.disabled) {
        return immutableInputBackgroundColor;
    } else {
        return mutableInputBackgroundColor;
    }
};

const getBorderColor = (props: TextInputProps) => {
    if (props.status !== Status.Undefined) {
        switch (props.status) {
            case Status.Invalid: return patternLabColors.status.error;
            case Status.Valid: return patternLabColors.status.success;
            default: return defaultBorderColor;
        }
    } else if (props.warning) {
        return warningBorderColor;
    } else if (props.readonly || props.disabled) {
        return statuslessImmutableBorderColor;
    } else {
        return defaultBorderColor;
    }
};

const getBorderColorOnHover = (props: TextInputProps) => {
    if (props.readonly || props.disabled) {
        return immutableBorderColorOnHover;
    } else {
        return defaultBorderColorOnHover;
    }
};

const getBorderColorOnFocus = (props: TextInputProps) => {
    if (props.readonly || props.disabled) {
        return immutableBorderColorOnFocus;
    } else {
        return patternLabColors.brand.primary;
    }
};

interface TextInputProps {
    disabled?: boolean;
    pattern?: string | RegExp;
    placeholder?: string;
    readonly?: boolean;
    status?: Status;
    warning?: boolean;
}

export const TextInput: StyledComponentClass<any, any> = styled.input.attrs<TextInputProps>({
    pattern: (props: TextInputProps) => props.pattern,
    placeholder: (props: TextInputProps) => props.placeholder,
    readOnly: (props: TextInputProps) => props.readonly,
    type: 'text',
}) `
    cursor: ${(props: TextInputProps) => props.disabled ? 'not-allowed' : 'text'};
    background-color: ${(props: TextInputProps) => getBackgroundColor(props)};
    border: 1px solid ${(props: TextInputProps) => getBorderColor(props)};
    border-radius: 2px;
    box-sizing: border-box;
    height: 35px;
    padding: 10px;
    width: 300px;

    :hover {
      border-color: ${(props: TextInputProps) => getBorderColorOnHover(props)}
    }

    :focus {
      border-color: ${(props: TextInputProps) => getBorderColorOnFocus(props)}
      outline: none;
    }
  `;
