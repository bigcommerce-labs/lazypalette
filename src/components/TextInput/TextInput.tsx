import React, {
  Component,
  ChangeEvent,
  ChangeEventHandler,
  FocusEvent,
  FocusEventHandler,
  SFC
} from 'react';
import uuid from 'uuid';

import { Status } from './Status';
import { Container, Label, Small, TextInput as StyledTextInput } from './styles';

interface LabelProps {
  children?: string;
  htmlFor?: string;
  note?: string;
  required?: boolean;
}

const TextInputLabel: SFC<LabelProps> = props => {
  const { children: label, htmlFor, note, required } = props;

  return (
    <Label htmlFor={htmlFor}>
      {label}
      {note &&
        <Small required={required!}>
          {note}
        </Small>
      }
    </Label>
  );
};

export interface TextInputState {
  status: Status;
  value: string | undefined;
}

export interface TextInputProps {
  defaultValue?: string | undefined;
  disabled?: boolean;
  inputId?: string;
  inset?: string;
  label?: string;
  note?: string;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  pattern?: string | RegExp;
  placeholder?: string;
  postfix?: string;
  prefix?: string;
  readonly?: boolean;
  required?: boolean;
  warning?: boolean;
}

export default class TextInput extends Component<TextInputProps, TextInputState> {
  static defaultProps = {
    disabled: false,
    inputId: uuid(),
    onBlur: () => undefined,
    onChange: () => undefined,
    readonly: false,
    required: false,
    warning: false,
  };

  readonly state: TextInputState = {
    status: Status.Undefined,
    value: '',
  };

  constructor(props: TextInputProps) {
    super(props);
    this.state = {
      status: Status.Undefined,
      value: props.defaultValue,
    };
  }

  componentWillMount() {
    if (this.props.pattern) {
      this.validate();
    }
  }

  render() {
    const {
      status,
    } = this.state;
    const {
      defaultValue,
      disabled,
      inputId,
      label,
      note = this.prepareNote(),
      pattern,
      placeholder,
      readonly,
      required,
      warning,
    } = this.props;

    return (
      <Container>
        <TextInputLabel htmlFor={inputId} note={note} required={required!}>
          {label}
        </TextInputLabel>
        <StyledTextInput
          defaultValue={defaultValue!}
          disabled={disabled!}
          id={inputId}
          onChange={this.onChange}
          onBlur={this.onBlur}
          pattern={pattern!}
          placeholder={placeholder!}
          readonly={readonly!}
          required={required!}
          status={status}
          warning={warning!}
        />
      </Container>
    );
  }

  private onBlur = (event: FocusEvent<HTMLInputElement>) => {
    this.props.onBlur!(event);

    if (this.props.pattern) {
      this.validate();
    }
  };

  private onChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({
      value: event.target.value,
    });

    this.props.onChange!(event);
  };

  private prepareNote() {
    if (this.props.note) {
      return this.props.note;
    } else if (this.props.required) {
      return 'Required*';
    } else {
      return undefined;
    }
  }

  private validate = () => {
    this.setState((prevState: Readonly<TextInputState>, props: TextInputProps) => {
      const regExp = props.pattern instanceof RegExp ? props.pattern : new RegExp(props.pattern!, 'u');

      return {
        status: regExp.test(prevState.value!) ? Status.Valid : Status.Invalid,
      };
    });
  };
}
