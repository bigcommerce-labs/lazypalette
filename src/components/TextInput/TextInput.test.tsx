import { shallow } from 'enzyme';
import React from 'react';

import TextInput from './TextInput';

const id = 'TextInputTestId';

it('renders', () => {
  const textInput = shallow(
    <TextInput inputId={id} label="Name your version" />
  );
  expect(textInput).toMatchSnapshot();
});

it('renders a valid state', () => {
  const textInput = shallow(
    <TextInput inputId={id} label="Name your version" defaultValue="abc" pattern="abc" />
  );
  expect(textInput).toMatchSnapshot();
});

it('renders an invalid state', () => {
  const textInput = shallow(
    <TextInput inputId={id} label="Name your version" defaultValue="adlkfj" pattern="abc" />
  );
  expect(textInput).toMatchSnapshot();
});
