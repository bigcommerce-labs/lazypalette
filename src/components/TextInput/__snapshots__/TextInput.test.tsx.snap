// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<styled.div>
  <TextInputLabel
    htmlFor="TextInputTestId"
    required={false}
  >
    Name your version
  </TextInputLabel>
  <styled.input
    disabled={false}
    id="TextInputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    readonly={false}
    required={false}
    status={2}
    warning={false}
  />
</styled.div>
`;

exports[`renders a valid state 1`] = `
<styled.div>
  <TextInputLabel
    htmlFor="TextInputTestId"
    required={false}
  >
    Name your version
  </TextInputLabel>
  <styled.input
    defaultValue="abc"
    disabled={false}
    id="TextInputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    pattern="abc"
    readonly={false}
    required={false}
    status={1}
    warning={false}
  />
</styled.div>
`;

exports[`renders an invalid state 1`] = `
<styled.div>
  <TextInputLabel
    htmlFor="TextInputTestId"
    required={false}
  >
    Name your version
  </TextInputLabel>
  <styled.input
    defaultValue="adlkfj"
    disabled={false}
    id="TextInputTestId"
    onBlur={[Function]}
    onChange={[Function]}
    pattern="abc"
    readonly={false}
    required={false}
    status={0}
    warning={false}
  />
</styled.div>
`;
